scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1500 @@
+    discrepancy corresponding to different collections of subsets. Some
+    versions take a root mean square difference over subsets instead of
+    a maximum.
+
+    A measure of uniformity is reasonable if it satisfies the following
+    criteria [1]_:
+
+    1. It is invariant under permuting factors and/or runs.
+    2. It is invariant under rotation of the coordinates.
+    3. It can measure not only uniformity of the sample over the hypercube,
+       but also the projection uniformity of the sample over non-empty
+       subset of lower dimension hypercubes.
+    4. There is some reasonable geometric meaning.
+    5. It is easy to compute.
+    6. It satisfies the Koksma-Hlawka-like inequality.
+    7. It is consistent with other criteria in experimental design.
+
+    Four methods are available:
+
+    * ``CD``: Centered Discrepancy - subspace involves a corner of the
+      hypercube
+    * ``WD``: Wrap-around Discrepancy - subspace can wrap around bounds
+    * ``MD``: Mixture Discrepancy - mix between CD/WD covering more criteria
+    * ``star``: Star L2-discrepancy - like CD BUT variant to rotation
+
+    Lastly, using ``iterative=True``, it is possible to compute the
+    discrepancy as if we had :math:`n+1` samples. This is useful if we want
+    to add a point to a sampling and check the candidate which would give the
+    lowest discrepancy. Then you could just update the discrepancy with
+    each candidate using `_update_discrepancy`. This method is faster than
+    computing the discrepancy for a large number of candidates.
+
+    References
+    ----------
+    .. [1] Fang et al. Design and modeling for computer experiments,
+       Computer Science and Data Analysis Series, 2006.
+    .. [2] Zhou Y.-D. et al. Mixture discrepancy for quasi-random point sets
+       Journal of Complexity, 29 (3-4) , pp. 283-301, 2013.
+    .. [3] T. T. Warnock. Computational investigations of low discrepancy point
+       sets, Applications of Number Theory to Numerical
+       Analysis, Academic Press, pp. 319-343, 1972.
+
+    Examples
+    --------
+    Calculate the quality of the sample using the discrepancy:
+
+    >>> from scipy.stats import qmc
+    >>> space = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+    >>> l_bounds = [0.5, 0.5]
+    >>> u_bounds = [6.5, 6.5]
+    >>> space = qmc.scale(space, l_bounds, u_bounds, reverse=True)
+    >>> space
+    array([[0.08333333, 0.41666667],
+           [0.25      , 0.91666667],
+           [0.41666667, 0.25      ],
+           [0.58333333, 0.75      ],
+           [0.75      , 0.08333333],
+           [0.91666667, 0.58333333]])
+    >>> qmc.discrepancy(space)
+    0.008142039609053464
+
+    """
+    sample = np.asarray(sample)
+
+    n, d = sample.shape
+
+    if iterative:
+        n += 1
+
+    if method == 'CD':
+        abs_ = abs(sample - 0.5)
+        disc1 = np.sum(np.prod(1 + 0.5 * abs_ - 0.5 * abs_ ** 2, axis=1))
+
+        prod_arr = 1
+        for i in range(d):
```

Comment:
I will assume you've tried to vectorize these or decided against it to save memory, so I won't try to eliminate any loops myself. Let's get this mathematically correct now and worry about optimizations in a future PR.

Reply:
I had a SO question for this one specifically https://stackoverflow.com/questions/44422370/why-is-numpy-slower-than-for-loop. I did lot of benchmark. Here is something I could retrieve:

```python
import numpy as np
import timeit
import openturns as ot
from numba import jit, njit


def discrepancy_initial(sample):
    disc1 = 0
    n_sample = len(sample)
    dim = sample.shape[1]

    for i in range(n_sample):
        prod = 1
        for item in sample[i]:
            sub = abs(item - 0.5)
            prod *= 1 + 0.5 * sub - 0.5 * sub ** 2
        disc1 += prod

    print(disc1)

    disc2 = 0
    for i in range(n_sample):
        for j in range(n_sample):
            prod = 1
            for k in range(dim):
                a = 0.5 * abs(sample[i, k] - 0.5)
                b = 0.5 * abs(sample[j, k] - 0.5)
                c = 0.5 * abs(sample[i, k] - sample[j, k])
                prod *= 1 + a + b - c
            disc2 += prod

    print(disc2)

    c2 = (13 / 12) ** dim - 2 / n_sample * disc1 + 1 / (n_sample ** 2) * disc2
    return np.sqrt(c2)


def discrepancy(sample):
    n_sample = len(sample)
    dim = sample.shape[1]

    abs_ = abs(sample - 0.5)
    disc1 = np.sum(np.prod(1 + 0.5 * abs_ - 0.5 * abs_ ** 2, axis=1))

    prod_arr = 1
    for i in range(dim):
        s0 = sample[:, i]
        prod_arr *= (1 +
                     0.5 * abs(s0[:, None] - 0.5) + 0.5 * abs(s0 - 0.5) -
                     0.5 * abs(s0[:, None] - s0))
    disc2 = prod_arr.sum()

    c2 = (13 / 12) ** dim - 2 / n_sample * disc1 + 1 / (n_sample ** 2) * disc2
    return np.sqrt(c2)

@jit
def discrepancy_numba(sample):
    n_sample = len(sample)
    dim = sample.shape[1]

    abs_ = abs(sample - 0.5)
    disc1 = np.sum(np.prod(1 + 0.5 * abs_ - 0.5 * abs_ ** 2, axis=1))

    prod_arr = 1
    for i in range(dim):
        s0 = sample[:, i]
        prod_arr *= (1 +
                     0.5 * abs(s0[:, None] - 0.5) + 0.5 * abs(s0 - 0.5) -
                     0.5 * abs(s0[:, None] - s0))
    disc2 = prod_arr.sum()

    c2 = (13 / 12) ** dim - 2 / n_sample * disc1 + 1 / (n_sample ** 2) * disc2
    return np.sqrt(c2)

@njit
def discrepancy_faster_numba(sample):
    disc1 = 0
    n_sample = len(sample)
    dim = sample.shape[1]

    for i in range(n_sample):
        prod = 1
        for item in sample[i]:
            sub = abs(item - 0.5)
            prod *= 1 + 0.5 * sub - 0.5 * sub ** 2
        disc1 += prod

    disc2 = 0
    for i in range(n_sample):
        for j in range(n_sample):
            prod = 1
            for k in range(dim):
                a = 0.5 * abs(sample[i, k] - 0.5)
                b = 0.5 * abs(sample[j, k] - 0.5)
                c = 0.5 * abs(sample[i, k] - sample[j, k])
                prod *= 1 + a + b - c
            disc2 += prod

    c2 = (13 / 12) ** dim - 2 / n_sample * disc1 + 1 / (n_sample ** 2) * disc2
    return np.sqrt(c2)

np.random.seed(0)
sample = np.random.random_sample((100, 2))
ot_sample = ot.Sample(sample)
# print(discrepancy(sample))
print(discrepancy_initial(sample))
# print(discrepancy_numba(sample))
# print(discrepancy_faster_numba(sample))
print(ot.SpaceFillingC2().evaluate(sample))

raise
# print('Initial time: ', timeit.repeat('discrepancy_initial(sample)', number=50, repeat=4, setup="from __main__ import discrepancy_initial, sample"))

vect_time = timeit.repeat('discrepancy(sample)', number=50, repeat=4, setup="from __main__ import discrepancy, sample")
numba_time = timeit.repeat('discrepancy_numba(sample)', number=50, repeat=4, setup="from __main__ import discrepancy_numba, sample")
fast_numba_time = timeit.repeat('discrepancy_faster_numba(sample)', number=50, repeat=4, setup="from __main__ import discrepancy_faster_numba, sample")
ot_time = timeit.repeat('ot.SpaceFillingC2().evaluate(ot_sample)', number=50, repeat=4, setup="from __main__ import ot_sample, ot")

print('vectorized time: ', vect_time)
print('numba time: ', numba_time)
print('Fast numba time: ', fast_numba_time)
print('OT time: ', ot_time)
print('Diff Vectorized / OT: ', np.array(vect_time) / np.array(ot_time))
```

Reply:
Great. Yeah, after writing my own implementations I'm happy with the speed of most of them (except L2-star).